File: Baekjoon/2579.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		Queue<Pos> q = new LinkedList<Pos>();

		Pos curPos;
		int[] arr;
		int[][] cache;
		int N, max =  -1;

		N = sc.nextInt();
		arr = new int[N+3];
		cache = new int[N+1][2];

		for (int i=1; i<=N; i++) {
			cache[i][0] =  -1;
			cache[i][1] =  -1;
			arr[i] = sc.nextInt();
		}

		q.add(new Pos(1, arr[1], 0));
		q.add(new Pos(2, arr[2], 0));

		while(!q.isEmpty()) {
			curPos = q.poll();
//			System.out.println("Log/ curPos.jumpTo: " + curPos.getJumpTo() + " curPos.getValue: " + curPos.getValue());

			if (curPos.getJumpTo() > N)
				continue;

			if (cache[curPos.getJumpTo()][curPos.getCountOne()] >= curPos.getValue())
				continue;
			else
				cache[curPos.getJumpTo()][curPos.getCountOne()] = curPos.getValue();

			if (curPos.getCountOne() == 1) {
				q.add(new Pos(curPos.getJumpTo()+2, curPos.getValue() + arr[curPos.getJumpTo()+2], 0));

			} else {
				q.add(new Pos(curPos.getJumpTo()+1, curPos.getValue() + arr[curPos.getJumpTo()+1], 1));
				q.add(new Pos(curPos.getJumpTo()+2, curPos.getValue() + arr[curPos.getJumpTo()+2], 0));
			}
		}

		q.clear();
		System.out.println(Max(cache[N][0], cache[N][1]));
	}
	public static int Max(int parm1, int parm2) {
		return parm1 > parm2 ? parm1:parm2;
	}
}

class Pos {
	private int jumpTo, value, countOne;

	Pos (int jumpTo, int value, int countOne) {
		this.jumpTo = jumpTo;
		this.value = value;
		this.countOne = countOne;
	}

	public int getJumpTo() {
		return jumpTo;
	}

	public int getValue() {
		return value;
	}

	public int getCountOne() {
		return countOne;
	}
}

File: Baekjoon/2579.cpp
#include <stdio.h>

#define Max(parm1, parm2)		((parm1>parm2?parm1:parm2))

int main() {
	int N, i;
	int arr[301];
	int dp[301];
	scanf ("%d", &N);

	for (i=1; i<=N; i++)
		scanf ("%d", &arr[i]);

	dp[0] = 0;
	dp[1] = arr[1];
	dp[2] = dp[1] + arr[2];
	dp[3] = Max(dp[1] + arr[3], arr[2] + arr[3]);
	for (i=4; i<=N; i++)
		dp[i] = Max(dp[i-2], dp[i-3] + arr[i-1]) + arr[i];
	printf ("%d", dp[N]);

//	for (i=0; i<=N; i++)
//		printf ("%d\n", dp[i]);

	return 0;
}
